File: DeleteandEarn.java
// Time Complexity : O(n+k)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int deleteAndEarn(int[] nums) {
        if (nums==null || nums.length==0){
            return 0;
        }
        int n = nums.length;
        //To find the max element
        int max=Integer.MIN_VALUE;
        for(int i =0;i<n;i++){
            max=Math.max(max,nums[i]);
        }
        int [] arr=new int[max+1];
        for(int i=0;i<n;i++){
            int curr=nums[i];
            arr[curr]+=curr;
        }
        // house robber
        int prev =arr[0];
        int curr=Math.max(arr[0],arr[1]);
        for(int i =2;i<arr.length;i++){
            int temp=curr;
            curr=Math.max(curr,arr[i]+prev);
            prev=temp;
        }
        return curr;
    }
}

File: DeleteAndEarn.cs
// time complexity - O(m+n) where n = length of nums and m = max value in nums
// space complexity - O(m) where m = max calue present in nums
public class Solution
{
    public int DeleteAndEarn(int[] nums)
    {
        int n = nums.Length;
        int maxVal = int.MinValue;
        for (int i = 0; i < n; i++)
        {
            maxVal = Math.Max(maxVal, nums[i]);
        }
        Console.WriteLine(maxVal);
        int[] numsVal = new int[maxVal];
        for (int i = 0; i < n; i++)
        {
            numsVal[nums[i] - 1] += nums[i];
        }
        if (numsVal.Length == 1)
        {
            return numsVal[0];
        }
        // int[] dp = new int[maxVal];
        // dp[0]= numsVal[0];
        // dp[1]= Math.Max(dp[0],numsVal[1]);
        int prev = numsVal[0];
        int curr = Math.Max(prev, numsVal[1]);

        for (int i = 2; i < maxVal; i++)
        {
            //dp[i] = Math.Max(dp[i-1],numsVal[i]+dp[i-2]);
            int tmp = curr;
            curr = Math.Max(curr, numsVal[i] + prev);
            prev = tmp;
        }
        //return dp[maxVal-1];
        return curr;
    }
}